{ts} whitelist filter pos errors

TomSmithCGAT opened the pull request:
- Adds options to whitelist to detect putative errors in the CBs above the knee threshols
- Adds tests to cover new options

See (#138) for motivation

IanSudbery commented on umi_tools/whitelist.py:
```diff
@@ -42,14 +42,39 @@
+
+Optional detection of putative error CBs above the threshold can be
+switched on with the error detection option:
+--ed-above-threshold
```
Why two different options? Why not `--ed-above-thershold=[discard|correct]` with a default setting of False/None?

TomSmithCGAT replied:
Yeah, that would be better. One other issue is what error threshold to allow for the error detection above the threshold. Right now, it re-uses the `--error-correct-threshold` which is still described as "Hamming distance for correction of barcodes to whitelist barcodes", e.g same error threshold correcting 1) below knee to above knee and 2) between the above knee barcodes. I assume this is the behaviour that users will desire but maybe there will be an edge case where different thresholds are desirable. For now, I'll just update the description for `--error-correct-threshold` to make this clearer.

TomSmithCGAT commented on umi_tools/umi_methods.py:
```diff
@@ -438,6 +440,70 @@ def getUserDefinedBarcodes(whitelist_tsv, getErrorCorrection=False):
+    '''
+
+    near_matches = []
+    comp_regex = regex.compile("(%s){e<=%i}" % (barcode, errors))
```
Yes, there must be. I'll look into this.

TomSmithCGAT commented:
As suggested, I've removed the `ed-resolution` option and just left the option `ed-above-threshold=[discard/correct]`, defaulted to `None`.

I've also implemented solution 1 above. E.g putative INDEL CBs are always discarded. The discarding/correcting behaviour is logged like so. Note the below is from the testing where we allow 3 errors in a 16-20 base CB in order to actually detect some "error" CBs. Hence the relatively large number of putative CB errors detected.

```
2019-02-06 15:52:02,654 INFO Top 414 cell barcodes passed the selected threshold
2019-02-06 15:52:02,991 INFO CBs above the knee corrected due to possible substitutions: 6
2019-02-06 15:52:02,991 INFO CBs above the knee discarded due to possible INDELs: 26
2019-02-06 15:52:02,991 INFO CBs above the knee discarded due to possible errors from multiple other CBs: 0


```

TomSmithCGAT commented:
As far as I'm aware, this branch is now good to merge?

IanSudbery commented:
okay, good to go

On Wed, 6 Feb 2019 at 15:55 Tom Smith <notifications@github.com> wrote:

> *@TomSmithCGAT* commented on this pull request.
> ------------------------------
>
> In umi_tools/umi_methods.py
> <https://github.com/CGATOxford/UMI-tools/pull/309#discussion_r254330427>:
>
> > @@ -438,6 +440,70 @@ def getUserDefinedBarcodes(whitelist_tsv, getErrorCorrection=False):
>      return set(cell_whitelist), false_to_true_map
>
>
> +def checkError(barcode, whitelist, errors=1):
> +    '''
> +    Check for errors (substitutions, insertions, deletions) between a barcode
> +    and a set of whitelist barcodes.
> +
> +    Returns the whitelist barcodes which match the input barcode
> +    allowing for errors. Returns as soon as two are identified.
> +    '''
> +
> +    near_matches = []
> +    comp_regex = regex.compile("(%s){e<=%i}" % (barcode, errors))
>
> Turns out regex is pretty quick. We can reduce run time for this step by
> approximately 20% but at a cost of increased dependencies. For 10,000 CBs
> above the knee, the current run-time is ~90s. For now, I'll leave this as
> it is
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CGATOxford/UMI-tools/pull/309#discussion_r254330427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFJFjrrTUvHtcEwqbBIFCJc_qksp284Cks5vKvr4gaJpZM4aRLdp>
> .
>
